### ISSUE TITLE
As a user, I want to build an Avatar, so I can show off my bad self.

### ISSUE THREAD

**Issue by owocki** [gitcoinco/web#813]
### User Story

[comment]: # (As a <user type>, I want to <task> so that <goal>.)
As a user, I want to build an Avatar, so I can show off my bad self.

### Why Is this Needed?

To make profiles on Gitcoin unique and peronsable... without being creepy

### Description

In the new FTUX flow ( https://github.com/gitcoinco/web/issues/441 ) please add a page that is an avatar builder.  This avatar builder will allow a user to select facial features (hats, glasses, earrings, beards, beard alt, mustaches, skin color, head shape, ears torso, mouth, head, nose, eyes, hair, hair color, and build an avatar

#### Grid
<img width="595" alt="screen shot 2018-04-05 at 3 03 51 pm" src="https://user-images.githubusercontent.com/513929/38391808-9ad9ef8c-38e2-11e8-9371-7877d4328681.png">

#### Library of Components
<img width="1608" alt="screen shot 2018-04-05 at 3 04 23 pm" src="https://user-images.githubusercontent.com/513929/38391823-a8d03aa6-38e2-11e8-8dec-bf769860b70a.png">

<img width="1475" alt="screen shot 2018-04-05 at 3 04 57 pm" src="https://user-images.githubusercontent.com/513929/38391854-c6d4eb82-38e2-11e8-9139-a1922972fb58.png">
<img width="1497" alt="screen shot 2018-04-05 at 3 04 52 pm" src="https://user-images.githubusercontent.com/513929/38391855-c6e6459e-38e2-11e8-923a-1078d59c2914.png">
<img width="1688" alt="screen shot 2018-04-05 at 3 04 47 pm" src="https://user-images.githubusercontent.com/513929/38391858-c6f9dfa0-38e2-11e8-9e28-e3c531e50803.png">

#### Final result

Please disregard the 3 icons overlaying the avatar. 

<img width="1660" alt="screen shot 2018-04-05 at 3 05 05 pm" src="https://user-images.githubusercontent.com/513929/38391841-bfc67eaa-38e2-11e8-9523-57fee7b4e380.png">

### Avatar builder specifications

<img width="1002" alt="screen shot 2018-04-07 at 12 49 20 pm" src="https://user-images.githubusercontent.com/238235/38457741-9f21c908-3a62-11e8-94fb-a411b2079f46.png">

<img width="1000" alt="screen shot 2018-04-07 at 12 50 24 pm" src="https://user-images.githubusercontent.com/238235/38457744-a3f41a3a-3a62-11e8-8902-d07a39fc7649.png">

<img width="995" alt="screen shot 2018-04-07 at 12 50 34 pm" src="https://user-images.githubusercontent.com/238235/38457748-aaeb3b34-3a62-11e8-8aeb-82de28c9baa7.png">

<img width="227" alt="screen shot 2018-04-07 at 12 50 43 pm" src="https://user-images.githubusercontent.com/238235/38457754-b0781c52-3a62-11e8-8989-b6d09cb8506a.png">

### Current Behavior
Users Github Profile image is Used

### Expected Behavior
* The avatar builder should show a live preview of the avatar as it's being assembled.
* The avatar builder should save and automatically sync to user's Gitcoin profile.
* The avatar builder should allow the user to save their avatar to their gitcoin profile

### Definition of Done
- All functionality exists and works as outlined in mockup and guidelines.
- Visual designs leverage Gitcoin [UI Kit styling](https://github.com/gitcoinco/creative/tree/master/Product/00_UI_Kit) (fonts, colors, etc) found in the Gitcoin Creative Repo.
- **Characteristics List** On the left show as square dots until user has made a selection of corresponding characteristic on the right, once selected, dot turns into a checkmark.
- **Avatar Work Area**  Please follow interaction hover and click states as shown in the visual guide
- Color bars will need to be reflective of the section the user is in. For instance skin tone colors will be different than accessory colors.
- A user will only be able to select one option from the following sections:
Face, Eyes, Nose, Mouth, Ears, Hair, Facial Hair, Clothing, and Background.
- A user may select multiple options from the following sections: Accessories.
- Once user completes the avatar, a light confirmation should display at the upper right hand corner of the avatar builder (placeholder is shown in the mockup).

### Assets
* [Avatar Builder Sketch File](https://github.com/gitcoinco/creative/tree/master/Product/Avatar)
* [Avatar Art Assets](https://github.com/gitcoinco/creative/tree/master/Product)
* [Gitcoin UI Kit ]((https://github.com/gitcoinco/creative/tree/master/Product/00_UI_Kit) 

### Additional Information
[comment]: # (Any other information that would be useful, content, screenshots, etc.)
Other Avatar Projects
* http://www.hexatar.com/
* https://github.com/alizohaib/avatar-maker

### Reviewers
@pixelantdesign @owocki

**Comment by thelostone-mc** [comment]
Are soul patches out of style ? :P

**Comment by gitcoinbot** [comment]
__This issue now has a funding of 0.65 ETH (239.97 USD @ $369.18/ETH) attached to it.__

 * If you would like to work on this issue you can claim it [here](https://gitcoin.co/issue/gitcoinco/web/813).
 * If you've completed this issue and want to claim the bounty you can do so [here](https://gitcoin.co/issue/gitcoinco/web/813)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $2962.31 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by vs77bb** [comment]
Love this.... The real question, will @thelostone-mc finally get an avatar that looks like him? 🤔

**Comment by gitcoinbot** [comment]
__Work has been started on the 0.65 ETH (257.92 USD @ $396.8/ETH) funding by__: 
 1. [@rickyplouis](https://gitcoin.co/profile/rickyplouis) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@owocki) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/issue/gitcoinco/web/813)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $2856.56 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by rickyplouis** [comment]
Sounds like a fun project that I would like to take on.

**Comment by PixelantDesign** [comment]
Please let us know if you have any questions @rickyplouis!

**Comment by owocki** [comment]
@rickyplouis glad to have you. 🙌 

let us know your plan of attack soon pls!

**Comment by rickyplouis** [comment]
@owocki Finished getting the environment setup locally with docker and integrated github. I'm still looking through the codebase and trying to determine where it would be best to build out the avatar  builder page. I was thinking of making it accessible by the user clicking on "change avatar" or something like that on the View Profile Page. Then once the on-boarding process is complete this page can be added into it later. Let me know what you think.

**Comment by owocki** [comment]
@rickyplouis we are building an onboarding flow ( https://github.com/gitcoinco/web/issues/441 ) where this will eventually live.

for now, lets keep the link to the page as lightweight as possible.  maybe just put it as a link [in the top nav](http://bits.owocki.com/0Q1d241p2f2i/Screen%20Shot%202018-04-10%20at%209.32.46%20AM.png)

**Comment by rickyplouis** [comment]
@owocki got it. I was also wondering if Gitcoin had a style guide or contribution guidelines that I should follow as I proceed?

**Comment by owocki** [comment]
style guide => https://github.com/gitcoinco/creative/tree/master/Brand
contribution guidelines => https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md

**Comment by rickyplouis** [comment]
@owocki awww thank you. I think I have everything I need to succeed. I'll get started right away. Give me a week or two.

**Comment by owocki** [comment]
@rickyplouis the team is checking in on this project right now.. how is it going?

**Comment by owocki** [comment]
I just removed ricky from the bounty.  If anyone else wants it, please say hi and let us know what your plan of attack is.

**Comment by Smona** [comment]
I have no experience with django specifically, but this seems like a really cool project and I'm happy to give it a shot. 

I'm setting up my local environment at the moment and finding the relevant files, and I've only seen reference to jquery in terms of frontend frameworks. Is any modern framework being used for the frontend? If not, would the team be amenable to selecting one (Vue, React, Preact, etc) to make available for its developers?

I can definitely implement this in jQuery, but it seems to be approaching the scope of interactivity where a framework could improve productivity. I'll keep getting set up and start creating the template in the meantime.

**Comment by gitcoinbot** [comment]
__Work has been started on the 0.65 ETH (385.54 USD @ $593.14/ETH) funding by__: 
 1. [@Smona](https://gitcoin.co/profile/Smona) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@owocki) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/issue/gitcoinco/web/813)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $5765.89 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by owocki** [comment]
its using jquery right now.. if you had your druthers what would it use?

**Comment by Smona** [comment]
I personally have extensive experience with Vue and love it. Vue is
currently on track to become the most used frontend framework iirc. If
download size is the highest priority, I've heard preact is the lightest,
and easy to work with if one knows react (which would also be a good
choice).

I don't want to be the person to entirely make the decision since I'm just
now getting involved. I think it's best if a project just sticks to one
framework, so the decision would affect future work on the frontend. Like I
said if the team wants to stick with jquery i can still tackle this, but if
the product owners think it's a worthwhile investment, my vote is for Vue.js

**Comment by Smona** [comment]
I've been making good progress on this:

![screen shot 2018-04-21 at 9 02 31 am](https://user-images.githubusercontent.com/7091399/39085013-c93015ca-4542-11e8-8acd-15851c63ab83.png)

Some questions & concerns:
- I can't find existing classes for the buttons matching the style in the mockup. Did I miss them, or should I write them?
- How will avatar configuration be persisted? As a configuration that's rendered on page load, or as an image?
- There are 3 kinds of color: skin, hair, and clothes. So far I've given these colors their own sections with square selectors: ![screen shot 2018-04-21 at 9 06 23 am](https://user-images.githubusercontent.com/7091399/39085060-50461ea6-4543-11e8-81a7-a1d831930bcc.png) 
but I imagine we could also display a persistent color picker as in the mockup and switch it between selecting these color categories depending on the open section. Thoughts?
- The positioning of the hair layer is incorrect; it's too far down.
- One of the hair colors was a duplicate so i removed it. Style 5 is missing color #8C3B28.
- More of a suggestion, but I'm a fantasy hair gal and it would be cool to have hair options like blue, purple, etc 👽
- Are the hover, selected, and remove-hover states in the mockup exclusive to the accessories page? The way I see it, you should be able to remove your hair, but not your face. And an 'add' hover state wouldn't make a lot of sense if you can only select one option in a section.

Based on progress so far, I think I could knock out the rest of this in a couple days ✨

**Comment by thelostone-mc** [comment]
@Smona

So I'm working on the ftux and I believe your work would go as one of the steps on the flow ( I'll leave an blank step for that ) 
PR : https://github.com/gitcoinco/web/pull/889

If you get the time, I'd like your feedback to see if anything needs to be changed ! 
If you need an extra pair oemf eyes to help review your code, just link the PR and I'd be happy to help :)

**Comment by PixelantDesign** [comment]
@Smona It’s looking great!

@thelostone-mc could you help with the button classes?

On hair positioning, feel free to adjust to where it makes sense.

In terms of configuration, I’m thinking image but might be better for @owocki or @mbeacom to answer that question.

I agree on the hair color, we can definitely expand the palette to include more cool colors.

I also agree on being able to select new hair and face shape but not completely remove.

Hope this helps!

**Comment by Smona** [comment]
@PixelantDesign good to hear :).

I feel it would make more sense to store avatar configuration as a collection of svg asset paths. These could be injected into an avatar element on the site and potentially cached. That solution seems to be the most efficient and simple.

I'll see what I can do for fixing the svgs. Should changes and new svgs be PR'd into https://github.com/gitcoinco/creative?

**Comment by owocki** [comment]
from emerson stone:

>  Kevin, let us know if you’ve run into any issues while testing the full set of avatar components we provided - just want to make sure that any new changes and instructions are reflected in the guide. 

@Smona i'll add you to an email thread with nina (from emerson stone -- our design agency) in case you want to see any changes here

> How will avatar configuration be persisted? As a configuration that's rendered on page load, or as an image?

it'd be great to be able to store a json blob of the avatar configuration off of the `Profile` object.  @mbeacom let me know if you strongly disagree.

it would also be nice to be able to reference the image itself with `<img src=''>` tags.. are you assembling the avatar on the frontend or backend right now?  theres a bunch of views in `dashboard/embed.py`that spit out images right now.. perhaps we could build another view that reads in the json blob for a given avatar and spits out their profile image.

> I'll see what I can do for fixing the svgs. Should changes and new svgs be PR'd into https://github.com/gitcoinco/creative?

sorry im new to the thread and not getting the context by looking above.  what needs fixing here?

**Comment by Smona** [comment]
@owocki some of the svg layers weren't lining up correctly. I've imported the existing svgs into sketch projects that made it easy to adjust all color variations of a layer in the same way or add new colors. They also export the resulting svgs with the correct filenames I've used in the implementation.

I imagine having these sketch files in the creative repo would make future updates to Avatar assets much easier.

I'm assembling the Avatar on the frontend by changing the src attributes of layered img tags. It would be trivial to send off a json object containing the static svg URLs. 

I'm not sure how one would go about converting all the layers into a single img on the back end.

**Comment by owocki** [comment]
@Smona -- it looks like @PixelantDesign added those assets to the creative repo [here](https://github.com/gitcoinco/creative/tree/master/Product/Avatar/SVG). feel free to PR back if anythings missing or changed!

> I'm assembling the Avatar on the frontend by changing the src attributes of layered img tags. It would be trivial to send off a json object containing the static svg URLs.

great! can you please plan to do so!?

> I'm not sure how one would go about converting all the layers into a single img on the back end.

got it.. hmm... 

i wonder if we could do this with https://github.com/astraw/svg_stack or https://github.com/btel/svg_utils

it'd be nice to only have to implement the avatar builder once (on the frontend in javascript OR in the backend) but if you are far along on the current implementation path i dont feel strongly you should change course.  carry on!

**Comment by Smona** [comment]
@PixelantDesign thanks!

Couple questions: 

Is the background color picker only displayed in the accessories section? 

Should hair be a required option? Seems like bald should be a choice.

**Comment by PixelantDesign** [comment]
Hi @Smona

The background is a separate option from the accessories (not nested under).

You are right that bald should be an option. At this point some of the head shapes are flat on top so we will need to fix that.

Could we set it up in a way so that hair isnt required?

**Comment by PixelantDesign** [comment]
Hi @smona. I’m not sure I understand your question. Feel free to hit me up on slack!

**Comment by Smona** [comment]
Progress update: all that's left to do for the frontend is add state styles for options and color picker. ETA tomorrow or the day after 🔨

**Comment by Smona** [comment]
The frontend is finished! @thelostone-mc and I are going to link up soon and complete this issue 🙂

**Comment by PixelantDesign** [comment]
Can’t wait to see it!

**Comment by gitcoinbot** [comment]
@Smona Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] auto removal (6 days)

**Comment by thelostone-mc** [comment]
@Smona could you leave a comment here telling gitcoin you are working on it ? ^_^

**Comment by PixelantDesign** [comment]
Hi Smona - thanks! sorry about the gitcoin bot.

**Comment by mbeacom** [comment]
@Smona Thanks for the update! Reaching out to you on slack!

**Comment by gitcoinbot** [comment]
@Smona Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=1) | [3 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=3) | [5 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=5) | [10 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=10) | [100 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@Smona due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/gitcoinco/web/813/223) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] warning (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=1) | [3 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=3) | [5 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=5) | [10 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=10) | [100 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=100)

**Comment by weberV** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 0.65 ETH (449.87 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/gitcoinco/web/813/223).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $3,594.16 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
@Smona Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=1) | [3 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=3) | [5 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=5) | [10 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=10) | [100 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@Smona due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/gitcoinco/web/813/223) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] warning (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=1) | [3 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=3) | [5 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=5) | [10 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=10) | [100 days](https://gitcoin.co/issue/gitcoinco/web/813/223?snooze=100)

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 0.65  ETH (286.98 USD @ $441.51/ETH) has been granted to @Smona for this issue. ⚡️ 

The sender had the following public comments: 
> thanks for your work on this. i  love my avatar!

Nice work @Smona! To redeem your tip, login to Gitcoin at https://gitcoin.co/explorer and select 'Claim Tip' from dropdown menu in the top right, or check your email for a link to the tip redemption page. 

 * $19462.83 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Expired** 

<hr>

__The funding of 0.65 ETH (286.98 USD @ $441.51/ETH) attached to this issue has been **cancelled** by the bounty submitter__

 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $19,864.34 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
